### PR TITLE
Tulip vault

### PR DESCRIPTION
**Pull request by Dappio-emerson**
Support V2 Orca vault.

### CODE REVIEW

File: src/tulip/index.ts [hunk 1]
```diff
@@ -85,62 +84,165 @@ export interface RaydiumVaultInfo extends VaultInfo {
...
-  degenUtilizationRate: BN;
-  loanToValueRatio: BN;
-  liquidationBonus: BN;
-  liquidationThreshold: BN;
-  minBorrowRate: BN;
-  optimalBorrowRate: BN;
-  degenBorrowRate: BN;
-  maxBorrowRate: BN;
-  fees: ReserveFees;
+export interface OrcaVault extends VaultInfo {
```
Comment (ironaddicteddog) [inline src/tulip/index.ts:87]: ~~`OrcaVault`~~ => `OrcaVaultInfo`

File: src/tulip/index.ts [hunk 2]
```diff
@@ -85,62 +84,165 @@ export interface RaydiumVaultInfo extends VaultInfo {
...
+    globalFarm: PublicKey;
+    farmTokenMint: PublicKey;
+    rewardTokenMint: PublicKey;
+    swapPoolMint: PublicKey;
+    tokenAMint: PublicKey;
+    tokenBMint: PublicKey;
+    swapMarkets: PublicKey[];
+  };
+}
+export interface OrcaDDVault extends VaultInfo {
```
Comment (ironaddicteddog) [inline src/tulip/index.ts:110]: ~~`OrcaDDVault`~~ => `OrcaDDVaultInfo`

File: src/tulip/infos.ts [hunk 3]
```diff
@@ -103,6 +172,26 @@ infos = class InstanceTulip {
       case RAYDIUM_VAULT_LAYOUT_SPAN:
         parseVault = this._parseRaydiumVault;
         break;
+      case ORCA_VAULT_LAYOUT_SPAN:
+        console.log("orca vault");
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:198]: Remove unused logs

File: src/tulip/infos.ts [hunk 4]
```diff
@@ -258,6 +347,139 @@ infos = class InstanceTulip {
       serumMarket,
     };
   }
+
+  private static _parseOrcaVault(data: any, vaultId: PublicKey): types.OrcaVault {
+    const decodeData = ORCA_VAULT_LAYOUT.decode(data);
+    const { base, farmData } = decodeData;
+
+    const vaultAccount = vaultV2Config.vaults.accounts.find((account) => account.orca?.account == vaultId.toString());
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:377]: What is `vaultV2Config` used for here? I see its imported from `ids` but it seems to keep `vaults.accounts`

Comment (Dappio-emerson) [reply src/tulip/infos.ts:375]: Add `feeCollector` from `vaultV2Config` this is an account need to use in ix, but didn't write in on-chain account.

Comment (ironaddicteddog) [reply src/tulip/infos.ts:375]: Oh yes I recalled. I think the word `vault` is kinda abused here. How about renaming:

- File name: ~~`ids/vault.ts`~~ => `ids/config.ts`
- Object name: ~~`vaultV2Config`~~ => `configV2`  ([ref](https://github.com/DappioWonderland/navigator/blob/tulip-vault/src/tulip/ids/vault.ts#L3))

In this case, the chain of invoking becomes: ~~`vaultV2Config.vaults.accounts`~~ => `configV2.vaults.accounts` which is simpler

File: src/tulip/index.ts [hunk 5]
```diff
@@ -85,62 +86,155 @@ export interface RaydiumVaultInfo extends VaultInfo {
...
-  feeReceiver: PublicKey;
-  oraclePubkey: PublicKey;
-  availableAmount: BN;
-  borrowedAmount: BN;
-  cumulativeBorrowRate: BN;
-  marketPrice: BN;
-  platformAmountWads: BN;
-  platformFee: BN;
-}
+// export interface LendingOptimizerVaultInfo extends VaultInfo {
```
Comment (ironaddicteddog) [inline src/tulip/index.ts:174]: Remove unused comments

File: src/tulip/index.ts [hunk 6]
```diff
@@ -85,62 +86,155 @@ export interface RaydiumVaultInfo extends VaultInfo {
...
+//   currentPlatformInformation: PublicKey;
+//   currentPlatformCount: PublicKey;
+//   lastRebaseSlot: BN;
+// }
 
-export interface LastUpdate {
-  lastUpdatedSlot: BN;
-  stale: boolean;
-}
+// export interface MultiDepositOptimizerVaultInfo extends VaultInfo {
```
Comment (ironaddicteddog) [inline src/tulip/index.ts:181]: Remove unused comments

File: src/tulip/infos.ts [hunk 7]
```diff
@@ -93,16 +101,61 @@ infos = class InstanceTulip {
...
+    const orcaFilters = [orcaSizeFilter];
+    const orcaConfig: GetProgramAccountsConfig = { filters: orcaFilters };
+
+    const orcaDDSizeFilter: DataSizeFilter = {
+      dataSize: ORCA_DD_VAULT_LAYOUT_SPAN,
+    };
+    const orcaDDFilters = [orcaDDSizeFilter];
+    const orcaDDConfig: GetProgramAccountsConfig = { filters: orcaDDFilters };
+
+    // const lendingOptimizerSizeFilter: DataSizeFilter = {
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:122]: Remove unused comments

File: src/tulip/infos.ts [hunk 8]
```diff
@@ -93,16 +101,61 @@ infos = class InstanceTulip {
...
+    // const multiDepositOptimizerSizeFilter: DataSizeFilter = {
+    //   dataSize: MULTI_DEPOSIT_OPTIMIZER_VAULT_LAYOUT_SPAN,
+    // };
+    // const multiDepositOptimizerFilters = [multiDepositOptimizerSizeFilter];
+    // const multiDepositOptimizerConfig: GetProgramAccountsConfig = { filters: multiDepositOptimizerFilters };
+
+    const raydiumVaults = paginate(await connection.getProgramAccounts(TULIP_VAULT_V2_PROGRAM_ID, raydiumConfig), page);
+    const orcaVaults = paginate(await connection.getProgramAccounts(TULIP_VAULT_V2_PROGRAM_ID, orcaConfig), page);
+    const orcaDDVaults = paginate(await connection.getProgramAccounts(TULIP_VAULT_V2_PROGRAM_ID, orcaDDConfig), page);
+    // const lendingOptimizerVaults = paginate(
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:137]: Remove unused comments

File: src/tulip/infos.ts [hunk 9]
```diff
@@ -93,16 +101,61 @@ infos = class InstanceTulip {
...
+    // );
+    // const multiDepositOptimizerVaults = paginate(
+    //   await connection.getProgramAccounts(TULIP_VAULT_V2_PROGRAM_ID, multiDepositOptimizerConfig),
+    //   page
+    // );
+    const vaultAccountInfos = [
+      ...raydiumVaults,
+      ...orcaVaults,
+      ...orcaDDVaults,
+      // ...lendingOptimizerVaults,
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:149]: Remove unused comments

File: src/tulip/infos.ts [hunk 10]
```diff
@@ -125,6 +179,18 @@ infos = class InstanceTulip {
       case RAYDIUM_VAULT_LAYOUT_SPAN:
         parseVault = this._parseRaydiumVault;
         break;
+      case ORCA_VAULT_LAYOUT_SPAN:
+        parseVault = this._parseOrcaVault;
+        break;
+      case ORCA_DD_VAULT_LAYOUT_SPAN:
+        parseVault = this._parseOrcaDDVault;
+        break;
+      // case LENDING_OPTIMIZER_VAULT_LAYOUT_SPAN:
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:188]: Remove unused comments

File: src/tulip/infos.ts [hunk 11]
```diff
@@ -280,8 +343,227 @@ infos = class InstanceTulip {
...
+      },
+      ddCompoundQueue,
+      ddCompoundQueueNonce,
+      ddConfigured,
+      ddWithdrawQueue,
+      ddWithdrawQueueNonce,
+    };
+  }
+
+  // private static _parseLendingOptimizerVault(data: any, vaultId: PublicKey): types.LendingOptimizerVaultInfo {
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:409]: Remove unused comments

File: src/tulip/infos.ts [hunk 12]
```diff
@@ -280,8 +343,227 @@ infos = class InstanceTulip {
...
+  ): Promise<types.VaultInfo[]> {
+    // Raydium:
+    //   tokenAccount: (pool coin, pool pc -> calc TVL), (reward A, reward B -> get Mint, then calc fee ATA)
+    // Orca:
+    //   tokenAccount: (pool coin, pool pc -> calc TVL)
+    //   mint: underlyingMint -> get authority (pool swap authority)
+    const tokenAccountKeys: PublicKey[] = [];
+    const mintKeys: PublicKey[] = [];
+    vaults.forEach((vault) => {
+      if (vault.type == types.VaultType.Raydium) {
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:455]: Use switch case?

File: src/tulip/infos.ts [hunk 13]
```diff
@@ -280,8 +343,227 @@ infos = class InstanceTulip {
...
+      tokenMap.set(t.pubkey.toBase58(), this._parseTokenAccount(t.pubkey, t.account?.data!))
+    );
+    const mintMap = new Map<string, Mint>();
+    (await getMultipleAccounts(connection, mintKeys)).forEach((m) =>
+      mintMap.set(m.pubkey.toBase58(), this._parseMint(m.pubkey, m.account?.data!))
+    );
+
+    const fetchedVaults: types.VaultInfo[] = [];
+    for (let vault of vaults) {
+      if (vault.type == types.VaultType.Raydium) {
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:488]: Use switch case?

File: src/tulip/index.ts [hunk 14]
```diff
@@ -155,5 +249,13 @@ export interface DepositorInfo extends IDepositorInfo {
...
   rewardPerSharePaid: BN;
-  extra_data_account: PublicKey;
+  extraDataAccount: PublicKey;
+}
+
+export enum VaultType {
+  Raydium,
+  Orca,
+  OrcaDD,
+  LendingOptimizer,
```
Comment (ironaddicteddog) [inline src/tulip/index.ts:259]: How about `Saber` and `Atrix`

Comment (Dappio-emerson) [reply src/tulip/index.ts:259]: Tulip team said they don't support Saber and Atrix any more.

Comment (ironaddicteddog) [reply src/tulip/index.ts:259]: Noted

File: src/tulip/infos.ts [hunk 15]
```diff
@@ -401,8 +705,33 @@ export class VaultInfoWrapper implements IVaultInfoWrapper {
   constructor(public vaultInfo: types.VaultInfo) {}
 
   getApr() {
-    // TODO
-    return 0;
+    const numberOfPeriods = 8760;
```
Comment (ironaddicteddog) [inline src/tulip/infos.ts:708]: Wonder how does `8760` come from? Perhaps add some descriptive comment here?

Comment (Dappio-emerson) [reply src/tulip/infos.ts]: Comment added.
this is yearly compound times = 24 * 365 = 8760

Comment (ironaddicteddog) [reply src/tulip/infos.ts]: Noted

**Pull request merged**